[BEAM-1899] Implementation of JStorm runner

bastiliu opened the pull request:
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

@kennknowles  @peihe  Could you please help to review the implementation of JStorm runner? 

Known issues:
- There are still two test suites which were failed(CountingSourceTest, GenerateSequenceTest). I am looking on them.
- SetState, SplittableParDo, AttemptedMetric, CommittedMetrics, and TestStream were not supported yet. So these parts are excluded in the integration test.

peihe commented:
Reviewed 14 of 63 files at r1.
Review status: 14 of 63 files reviewed at latest revision, 35 unresolved discussions, some commit checks failed.

---

*[examples/java8/pom.xml, line 46 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnEUAmcZ3QdWzSP-B5:-KpnEUAmcZ3QdWzSP-B6:b6n2paf) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/examples/java8/pom.xml#L46)):*
> ```XML
>       A default profile that includes optional dependencies
>       on all of our runners. This is aimed at making it very
>       easy for users to run examples with any runner without
> ```

revert this file.

---

*[runners/jstorm/pom.xml, line 58 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnEgYS77euoNq1Lmvv:-KpnEgYS77euoNq1Lmvw:b-hay66i) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/pom.xml#L58)):*
> ```XML
>                   <!--<includes>
>                     <include>org.apache.beam.sdk.transforms.CombineTest.java</include>
>                   </includes>-->
> ```

revert?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 33 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnEs_a9fArGOkK3a5E:-KpnEs_a9fArGOkK3a5F:bbuihn7) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L33)):*
> ```Java
> 
>   @Description("Indicate if the topology is running on local machine or distributed cluster")
>   @Default.Boolean(false)
> ```

default to true to make it easier for users to get start?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 39 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFVrWGCJOM_s8G65T:-KpnFVrXBpH8C8kV2Eot:b-hsc6qb) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L39)):*
> ```Java
>   @Description("Executing time(sec) of topology on local mode. Default is 1min.")
>   @Default.Long(60)
>   Long getLocalModeExecuteTime();
> ```

getLocalModeExecuteTimeSec()

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 47 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFfvu1eCHOrxXsTjs:-KpnFfvu1eCHOrxXsTjt:b-wruxx5) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L47)):*
> ```Java
>   void setWorkerNumber(Integer number);
> 
>   @Description("Global parallelism number of a component")
> ```

is this for a component or for a topology?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 49 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFKAtEy4g5sBKVnBK:-KpnFKAtEy4g5sBKVnBL:b-mfbp8f) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L49)):*
> ```Java
>   @Description("Global parallelism number of a component")
>   @Default.Integer(1)
>   Integer getParallelismNumber();
> ```

nit: getParallelism()

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 64 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFwlwv9YNgBJL_ETz:-KpnFwlwv9YNgBJL_EU-:bxezwzh) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L64)):*
> ```Java
>   @Description("Parallelism number of a specified composite PTransform")
>   @Default.InstanceFactory(DefaultMapValueFactory.class)
>   Map getParallelismNumMap();
> ```

nit: getParallelismMap()

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 70 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnG5Byur07qcT29Ocw:-KpnG5Byur07qcT29Ocx:by144h8) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L70)):*
> ```Java
>    * Default value factory for topology configuration of JStorm.
>    */
>   class DefaultMapValueFactory implements DefaultValueFactory<Map> {
> ```

Is this to avoid NPE? Do we need it?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java, line 76 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnGnEvcv1AqJ4Wj-E1:-KpnGnEwg7aHr7o70UbG:benxvto) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java#L76)):*
> ```Java
>   public static JStormRunner fromOptions(PipelineOptions options) {
>     JStormPipelineOptions pipelineOptions = PipelineOptionsValidator.validate(
>         JStormPipelineOptions.class, options);
> ```

nit: put PipelineOptionsValidator.validate(JStormPipelineOptions.class, options); in the second line

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java, line 83 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnGcDt_wxNhyocmgAb:-KpnGcDt_wxNhyocmgAc:b-lqmk6z) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java#L83)):*
> ```Java
>    * convert pipeline options to storm configuration format.
>    * @param options
>    * @return
> ```

remove or fix this javadoc

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java, line 99 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnH0Um1aDYwu-mSU5_:-KpnH0Um1aDYwu-mSU5a:b-jk3e4v) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java#L99)):*
> ```Java
>     // Setup config for runtime env
>     config.put("worker.external", "beam");
>     config.put("topology.acker.executors", 0);
> ```

why set topology.acker.executors to 0, add comments?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 65 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIARz1uhCCA3sXaAp:-KpnIARz1uhCCA3sXaAq:baqxodb) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L65)):*
> ```Java
>   }
> 
>   private static class LocalStormPipelineResult extends JStormRunnerResult {
> ```

-> LocalJStormPipelineResult

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 81 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIIdt32knt4vP3zmS:-KpnIIdu7qnQkAaiG0FK:b-l6mwxj) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L81)):*
> ```Java
>     @Override
>     public State cancel() throws IOException {
>       //localCluster.deactivate(getTopologyName());
> ```

remove

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 95 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIfPcEpRgNYrnzkJF:-KpnIfPcEpRgNYrnzkJG:b-amvwii) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L95)):*
> ```Java
>     @Override
>     public State waitUntilFinish() {
>       JStormUtils.sleepMs(localModeExecuteTimeSecs * 1000);
> ```

call waitUntilFinish(Duration duration) from here with localModeExecuteTimeSecs.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 105 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIRgp8oy9jtaMwFve:-KpnIRgp8oy9jtaMwFvf:bhlv2mw) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L105)):*
> ```Java
>     @Override
>     public MetricResults metrics() {
>       return null;
> ```

revert to throw

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java, line 59 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIopPtw5FgQ__Lpn2:-KpnIopPtw5FgQ__Lpn3:b-l6mwxj) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java#L59)):*
> ```Java
>     this.options = options;
>     Map conf = Maps.newHashMap();
>     //conf.put(ConfigExtension.KV_STORE_TYPE, KvStoreManagerFactory.KvStoreType.memory.toString());
> ```

remove

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java, line 77 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJ3QzaWZek-FlOJr2:-KpnJ3QzaWZek-FlOJr3:b-ttjljy) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java#L77)):*
> ```Java
> 
>       int maxTimeoutSec = numberOfAssertions > 0 ? 20 : 5;
>       for (int waitTime = 0; waitTime <= maxTimeoutSec * 1000; ) {
> ```

define those numbers as private static final variables?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java, line 90 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJDtEXcB9UVWPSoaT:-KpnJDtEXcB9UVWPSoaU:b-7ropln) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java#L90)):*
> ```Java
>         } else {
>           JStormUtils.sleepMs(500);
>           waitTime += 500;
> ```

ditto

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java, line 27 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJVlgkgpM8ODzBjTu:-KpnJVlgkgpM8ODzBjTv:bumwfes) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java#L27)):*
> ```Java
> 
> /**
>  * Translates a {@link Read.Bounded} into a Storm spout.
> ```

JStorm

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java, line 29 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJTVb6_THmfGitDPK:-KpnJTVb6_THmfGitDPL:b-vabat8) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java#L29)):*
> ```Java
>  * Translates a {@link Read.Bounded} into a Storm spout.
>  *
>  * @param <T>
> ```

remove @param

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java, line 33 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnK8ZCJuAUBey6spTt:-KpnK8ZCJuAUBey6spTu:b5fubzl) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java#L33)):*
> ```Java
>   void init(ExecutorContext context);
> 
>   <T> void process(TupleTag<T> tag, WindowedValue<T> elem);
> ```

javadoc

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java, line 35 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnK9bUvwLcnEu6bKvm:-KpnK9bUvwLcnEu6bKvn:b5fubzl) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java#L35)):*
> ```Java
>   <T> void process(TupleTag<T> tag, WindowedValue<T> elem);
> 
>   void cleanup();
> ```

javadoc

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ExecutorContext.java, line 40 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnKHMI4LPpC2MgjgMH:-KpnKHMI4LPpC2MgjgMI:bhnuf67) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ExecutorContext.java#L40)):*
> ```Java
>   public abstract ExecutorsBolt getExecutorsBolt();
> 
>   public abstract IKvStoreManager getKvStoreManager();
> ```

could we remove public from above?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/FlattenExecutor.java, line 29 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnK_eLKedjxY3Fi6N7:-KpnK_eLKedjxY3Fi6N8:b-vabat8) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/FlattenExecutor.java#L29)):*
> ```Java
> /**
>  * JStorm {@link Executor} for {@link org.apache.beam.sdk.transforms.Flatten}.
>  * @param <InputT>
> ```

remove @param

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByKeyTranslator.java, line 33 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnKlRCkrDUOmg_Llcn:-KpnKlRCkrDUOmg_Llco:b-9g5aiy) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByKeyTranslator.java#L33)):*
> ```Java
>  * Translates a {@link GroupByKey} to a JStorm {@link GroupByWindowExecutor}.
>  * @param <K>
>  * @param <V>
> ```

remove param

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByWindowExecutor.java, line 55 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnKqLvH26dWG55oac2:-KpnKqLvH26dWG55oac3:b-8s3ubn) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByWindowExecutor.java#L55)):*
> ```Java
>  * JStorm {@link Executor} for {@link org.apache.beam.sdk.transforms.GroupByKey}.
>  * @param <K>
>  * @param <V>
> ```

remove params

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/MultiOutputDoFnExecutor.java, line 36 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnLCMYqC2y3cA9ILeW:-KpnLCMYqC2y3cA9ILeX:b-8s3ubn) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/MultiOutputDoFnExecutor.java#L36)):*
> ```Java
>  * JStorm {@link Executor} for {@link DoFn} with multi-output.
>  * @param <InputT>
>  * @param <OutputT>
> ```

remove params

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerService.java, line 37 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnL_3rZrs1emqrNVtL:-KpnL_3rZrs1emqrNVtM:b-tl4hjl) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerService.java#L37)):*
> ```Java
>    * @param inputWatermark
>    * @return new watermark if any timer is triggered during the update of watermark, otherwise 0
>    */
> ```

update javadocs

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerServiceImpl.java, line 89 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnLzNNGHrY3ioRd6ii:-KpnLzNNGHrY3ioRd6ij:b-azs21l) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerServiceImpl.java#L89)):*
> ```Java
>       }
>     }
>     return 0;
> ```

how do we use the return value?
Is that right to return 0?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TransformTranslator.java, line 43 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMBOYNJD5YJDlVe_U:-KpnMBOYNJD5YJDlVe_V:b-c97nd8) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TransformTranslator.java#L43)):*
> ```Java
>      * Default translator.
>      * @param <T1>
>      */
> ```

javadoc is off. Remove param, and say more about Default translator.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TranslationContext.java, line 202 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMOkGUJRak6wNl8dY:-KpnMOkGUJRak6wNl8dZ:b-hfpv1m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TranslationContext.java#L202)):*
> ```Java
> 
>       // filter all connections inside bolt
>       //if (!bolt.getOutputTags().contains(tag)) {
> ```

remove?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/UnboundedSourceSpout.java, line 62 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMfSXCAT9y-x0klRV:-KpnMfSXCAT9y-x0klRW:b3xavp7) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/UnboundedSourceSpout.java#L62)):*
> ```Java
>   private KryoSerializer<WindowedValue> serializer;
> 
>   private long lastWaterMark = 0L;
> ```

BoundedWindow.TIMESTAMP_MIN_VALUE.getMillis()?

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/WindowAssignExecutor.java, line 35 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMwOU6VlUCapNB-mx:-KpnMwOU6VlUCapNB-my:b-8s3ubn) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/WindowAssignExecutor.java#L35)):*
> ```Java
>  * JStorm {@link Executor} for {@link org.apache.beam.sdk.transforms.windowing.Window.Assign}.
>  * @param <T>
>  * @param <W>
> ```

remove params

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java, line 236 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnNCKwYWHAT7HsZkJl:-KpnNCKwYWHAT7HsZkJm:bpl1b4g) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java#L236)):*
> ```Java
>     @Override
>     public Iterable<T> apply(Iterable<WindowedValue<T>> contents) {
>       return ImmutableList.copyOf(
> ```

Need to figure out how to get rid of this hack.

---

*[sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/CombineTest.java, line 465 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnNO4W2oLn2vG9xIey:-KpnNO4W2oLn2vG9xIez:b-6nr0jd) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/CombineTest.java#L465)):*
> ```Java
>   public void testAccumulatingCombine() {
>     runTestAccumulatingCombine(
>         TABLE, 4.0, Arrays.asList(KV.of("a", 2.0), KV.of("b", 7.0)));
> ```

revert this file

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3624)*
<!-- Sent from Reviewable.io -->

peihe commented on runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ParDoBoundTranslator.java:
```diff
@@ -0,0 +1,108 @@
+/*
```
I think we can remove this class because SingleOutput will be translated to MultiOutput by default.

kennknowles replied:
<!--thread_id:cc_130004480_t; commit:6d1d2030507abb560abfea721679f2f92734618c; resolved:0-->
<!--section:context-quote-->
> **peihe** wrote:
> I think we can remove this class because SingleOutput will be translated to MultiOutput by default.

<!--section:body-->
This is true. Only multi output ParDo is a real Beam primitive. You can optionally use a replacement to have a specialized single output ParDo if that gains performance or makes some UI easier to read.

kennknowles commented:
R: @kennknowles

I'll take a look through this, too. I will focus on getting the module and branch set up right, as there will be time to continue to improve it before moving it to `master`.

kennknowles commented on examples/java8/pom.xml:
```diff
@@ -68,7 +69,7 @@
           <scope>runtime</scope>
         </dependency>
         <!--
-          Apex depends on httpclient version 4.3.5, project has a transitive dependency to httpclient 4.0.1 from
```
<!--new_thread; commit:0a05de365ddaff9d1c570423ab7ea527b5bf77ae; resolved:0-->
Accidental deletion?

bastiliu replied:
Yes. It is a merge problem. I will revert it.

kennknowles commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ReshuffleTest.java:
```diff
@@ -70,9 +73,16 @@
-  private static final List<KV<String, Iterable<Integer>>> GROUPED_TESTABLE_KVS = ImmutableList.of(
-        KV.of("k1", (Iterable<Integer>) ImmutableList.of(3)),
-        KV.of("k2", (Iterable<Integer>) ImmutableList.of(4)));
+  private static class AssertThatHasExpectedContents
```
<!--new_thread; commit:9f415d4d04ecbb6a985de9e9a27f27e427694877; resolved:0-->
I think you can just switch this to be a list of matchers so you can say

PAssert.that(collection).containsInAnyOrder(isKv(is("k1"), containsInAnyOrder(3)), ...)

and so on.

bastiliu replied:
Addressed.

kennknowles commented on runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/JStormStateInternals.java:
```diff
@@ -0,0 +1,831 @@
+
+    @Override
+    public CombiningState<InputT, AccumT, OutputT> readLater() {
+      // TODO: support prefetch.
```
<!--new_thread; commit:13d5a9c5b6f60fd83d944812acd8500e37d17f48; resolved:0-->
I think it would be good to file JIRA for this (and other) TODOs. They may be lost in the large merge of the PR.

bastiliu replied:
OK, I will create a JIRA.

kennknowles commented on runners/jstorm/pom.xml:
```diff
@@ -53,18 +53,29 @@
                 </goals>
                 <configuration>
+                  <!--<includes>
+                    <include>org.apache.beam.sdk.transforms.CombineTest.java</include>
```
<!--new_thread; commit:d8cb47c12a014273ba68adce0042f23f9e3e529e; resolved:0-->
Delete this?

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java:
```diff
@@ -233,7 +233,7 @@ public T apply(Iterable<WindowedValue<T>> contents) {
     @Override
     public Iterable<T> apply(Iterable<WindowedValue<T>> contents) {
-      return Iterables.unmodifiableIterable(
+      return ImmutableList.copyOf(
```
<!--new_thread; commit:a2ad4c5858521a9f42105f87fa3401fbab5bff18; resolved:0-->
Can't do this - the contents might be an entire PCollection streamed in.

bastiliu replied:
Reverted. It is just a temporary workaround for a deserialization problem which has been fixed now.

kennknowles commented on sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java:
```diff
@@ -112,23 +112,7 @@ public void testCreationOfPipelineOptionsFromReallyVerboselyNamedTestCase() thro
-
-    private TestPipeline nestedMethod() {
-      return TestPipeline.create();
+          pipeline.toString());
```
<!--new_thread; commit:fc00631b35848d0801535b00ee91c13d9a9d58c6; resolved:0-->
Why this change?

bastiliu replied:
Please refer to JIRA " [BEAM-2630] TestPipeline: construct job/app names based on Description in junit TestRule."

kennknowles commented on runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TranslatorRegistry.java:
```diff
@@ -0,0 +1,58 @@
+class TranslatorRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(TranslatorRegistry.class);
+
+  private static final Map<Class<? extends PTransform>, TransformTranslator> TRANSLATORS =
```
<!--new_thread; commit:833945697e16b161adcffcaa56d02a991a85b7a0; resolved:1-->
For cross-language portability, you will want to key this on URN. It does not have to happen right now.

bastiliu replied:
OK, I will file a JIRA for support of cross-language.

kennknowles commented on runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/SerializedPipelineOptions.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Encapsulates the PipelineOptions in serialized form to ship them to the cluster.
+ */
+public class SerializedPipelineOptions implements Serializable {
```
<!--new_thread; commit:833945697e16b161adcffcaa56d02a991a85b7a0; resolved:1-->
I think @jkff recently made a shared utility for doing this. Maybe it works for you, or not, either way is fine.

jkff replied:
Yup, please use https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java

kennknowles commented on runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java:
```diff
@@ -17,47 +17,92 @@
     @Override
     public MetricResults metrics() {
-        throw new UnsupportedOperationException("This method is not yet supported.");
+      return null;
```
<!--new_thread; commit:dec02cd228da604a68e5779a5c10935881fd37a1; resolved:0-->
Should still throw, yes? Still unsupported, clearly.

bastiliu commented:
Review status: 10 of 59 files reviewed at latest revision, 36 unresolved discussions, some commit checks failed.

---

*[examples/java8/pom.xml, line 46 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnEUAmcZ3QdWzSP-B5:-Kr9mJAkNF4In_79NMNH:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/examples/java8/pom.xml#L46)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

revert this file.
</blockquote></details>

Addressed

---

*[runners/jstorm/pom.xml, line 58 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnEgYS77euoNq1Lmvv:-Kr92PJutbDUEb1k4LCE:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/pom.xml#L58)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

revert?
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 33 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnEs_a9fArGOkK3a5E:-Kr96tVmVzPnDRyjx0UE:brzii2m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L33)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

default to true to make it easier for users to get start?
</blockquote></details>

Since local mode job would fail in distributed cluster, it is better to set false as default config here. For local testing user, TestJStormRunner can be used,  which includes some other configuration for testing.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 39 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFVrWGCJOM_s8G65T:-Kr96DHAJ0R8-gvQOQSf:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L39)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

getLocalModeExecuteTimeSec()
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 47 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFfvu1eCHOrxXsTjs:-Kr97QSMrTuK7nJ0F9oI:b2nhswy) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L47)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

is this for a component or for a topology?
</blockquote></details>

Yes

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 49 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFKAtEy4g5sBKVnBK:-Kr97RKbxiC9DVdD9oQE:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L49)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

nit: getParallelism()
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 64 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnFwlwv9YNgBJL_ETz:-Kr97lKoc3e7y9wtYERZ:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L64)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

nit: getParallelismMap()
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java, line 70 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnG5Byur07qcT29Ocw:-Kr98-tuA54FgghjyK6h:b-53pxf0) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormPipelineOptions.java#L70)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

Is this to avoid NPE? Do we need it?
</blockquote></details>

Yes. But if there is already a definition of Default Map in Beam, then we can remove it.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java, line 76 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnGnEvcv1AqJ4Wj-E1:-Kr98vtWOVMJOZFCpTFi:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java#L76)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

nit: put PipelineOptionsValidator.validate(JStormPipelineOptions.class, options); in the second line
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java, line 83 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnGcDt_wxNhyocmgAb:-Kr992m6tPfDw_eqxIbK:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java#L83)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove or fix this javadoc
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java, line 99 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnH0Um1aDYwu-mSU5_:-Kr9947QSFfMK0Mlx9gm:br39uqj) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunner.java#L99)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

why set topology.acker.executors to 0, add comments?
</blockquote></details>

We don't need acker for "at least once" for JStorm Beam job. For saving unnecessary cost of task, we set executors number of acker to 0. Comments were added.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 65 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIARz1uhCCA3sXaAp:-Kr99eHr5quZDOsgAfVv:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L65)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

-> LocalJStormPipelineResult
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 81 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIIdt32knt4vP3zmS:-Kr9CLx8EiPYiKalM4tL:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L81)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 95 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIfPcEpRgNYrnzkJF:-Kr9DQP3YxdIGWCwvhVL:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L95)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

call waitUntilFinish(Duration duration) from here with localModeExecuteTimeSecs.
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java, line 105 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIRgp8oy9jtaMwFve:-Kr9Dmuajc5Em34CDRyq:bjcwa0u) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/JStormRunnerResult.java#L105)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

revert to throw
</blockquote></details>

throw UnsupportedOperationException

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java, line 59 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnIopPtw5FgQ__Lpn2:-Kr9DwmCjIEGyP4XvYsd:bl6to4m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java#L59)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove
</blockquote></details>

Changed to comments here.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java, line 77 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJ3QzaWZek-FlOJr2:-Kr9EjnelnuDjQBD7Q-S:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java#L77)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

define those numbers as private static final variables?
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java, line 90 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJDtEXcB9UVWPSoaT:-Kr9FkGmmrVrhzzh7c8M:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/TestJStormRunner.java#L90)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java, line 27 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJVlgkgpM8ODzBjTu:-Kr9GIkw0JoXU7s1eHFl:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java#L27)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

JStorm
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java, line 29 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnJTVb6_THmfGitDPK:-Kr9GMrdVLtR3bGsHV6d:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/BoundedSourceTranslator.java#L29)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove @param
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java, line 33 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnK8ZCJuAUBey6spTt:-Kr9GdU0oqLKyzS_Hmvx:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java#L33)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

javadoc
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java, line 35 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnK9bUvwLcnEu6bKvm:-Kr9GeENoagG7U194kwX:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/Executor.java#L35)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

javadoc
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ExecutorContext.java, line 40 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnKHMI4LPpC2MgjgMH:-Kr9HGzLWcDxXIJWGBHB:b-hmla4u) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ExecutorContext.java#L40)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

could we remove public from above?
</blockquote></details>

No, context information is required when runtime.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/FlattenExecutor.java, line 29 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnK_eLKedjxY3Fi6N7:-Kr9OlW_Y1ag-zZNbSkS:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/FlattenExecutor.java#L29)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove @param
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByKeyTranslator.java, line 33 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnKlRCkrDUOmg_Llcn:-Kr9OrdIgMSkFiyAMmOW:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByKeyTranslator.java#L33)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove param
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByWindowExecutor.java, line 55 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnKqLvH26dWG55oac2:-Kr9OzhBG6zmxE445rG_:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/GroupByWindowExecutor.java#L55)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove params
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/MultiOutputDoFnExecutor.java, line 36 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnLCMYqC2y3cA9ILeW:-Kr9P5IX_BvSvevdlaHf:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/MultiOutputDoFnExecutor.java#L36)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove params
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ParDoBoundTranslator.java, line 1 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnDXrTfgmPpo0jNvWY-r1-1:-Kr9fbNoXqdDsc7iYdAI:bfbjc5n) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/ParDoBoundTranslator.java#L1)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

<!--thread_id:cc_130004480_t; commit:6d1d2030507abb560abfea721679f2f92734618c; resolved:0-->
<!--section:context-quote-->
> **peihe** wrote:
> I think we can remove this class because SingleOutput will be translated to MultiOutput by default.

<!--section:body-->
This is true. Only multi output ParDo is a real Beam primitive. You can optionally use a replacement to have a specialized single output ParDo if that gains performance or makes some UI easier to read.
</blockquote></details>

OK. I will remove it.

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerService.java, line 37 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnL_3rZrs1emqrNVtL:-Kr9gF7BOjAjITp5gBbM:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerService.java#L37)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

update javadocs
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerServiceImpl.java, line 89 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnLzNNGHrY3ioRd6ii:-Kr9ityaVrWzHnwETSDP:b-j6rf8o) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TimerServiceImpl.java#L89)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

how do we use the return value?
Is that right to return 0?
</blockquote></details>

The returned non-zero WM here is used to fire the timers which are earlier than this time while zero means none timer will be fired. 

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TransformTranslator.java, line 43 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMBOYNJD5YJDlVe_U:-Kr9jyx4qXIg98LlouyS:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TransformTranslator.java#L43)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

javadoc is off. Remove param, and say more about Default translator.
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TranslationContext.java, line 202 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMOkGUJRak6wNl8dY:-Kr9l_hA6joWTQ-368f2:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/TranslationContext.java#L202)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove?
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/UnboundedSourceSpout.java, line 62 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMfSXCAT9y-x0klRV:-Kr9ll1UAO1XAeFQ1cS7:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/UnboundedSourceSpout.java#L62)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

BoundedWindow.TIMESTAMP_MIN_VALUE.getMillis()?
</blockquote></details>

Addressed

---

*[runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/WindowAssignExecutor.java, line 35 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnMwOU6VlUCapNB-mx:-Kr9m3ciTwrsXXN_q3_G:bx3iq6m) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/WindowAssignExecutor.java#L35)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

remove params
</blockquote></details>

Addressed

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java, line 236 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnNCKwYWHAT7HsZkJl:-KrTTbNt7RQBgL8RGhEN:b49sqmf) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java#L236)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

Need to figure out how to get rid of this hack.
</blockquote></details>

It is a temporary workaround for the serialization problem of Kryo.  Now, it can be reverted.  The serialization problem was fixed.

---

*[sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/CombineTest.java, line 465 at r1](https://reviewable.io:443/reviews/apache/beam/3624#-KpnNO4W2oLn2vG9xIey:-Kr9PJqKQ1l-ggDRpfdY:bvljs2o) ([raw file](https://github.com/apache/beam/blob/6d1d2030507abb560abfea721679f2f92734618c/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/CombineTest.java#L465)):*
<details><summary><i>Previously, peihe wrote…</i></summary><blockquote>

revert this file
</blockquote></details>

Addressed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/3624)*
<!-- Sent from Reviewable.io -->

bastiliu commented:
@peihe  @kennknowles  I have addressed your comments. Could you please help to check?

bastiliu commented:
@kennknowles  Thanks for your review. I have sent out the email of ICLA today. Any response on this, I will let you know.

bastiliu commented:
@kennknowles  ICLA was done.

bastiliu commented:
The testing failure of sdk python module blocked the validation of JStorm runner, So I comment out this module in last commit. After the validation test of JStorm runner, this commit will be reverted.

kennknowles commented:
Some findbugs errors are remaining.

bastiliu commented:
Findbugs were Fixed.

peihe commented:
I appended few fixups, and Jenkins passed in https://github.com/apache/beam/pull/3734

@kennknowles, I can merge, if this PR looks good to you.

peihe commented:
This PR is merged in https://github.com/apache/beam/pull/3734
We can close it now.

Thanks, guys!

bastiliu commented:
Closed.
